EC2SecurityGroup: add TagSpecifications parameter when Tags present in CloudFormation template

jcollado opened the pull request:
<!-- Please refer to the contribution guidelines before raising a PR: https://github.com/localstack/localstack/blob/master/docs/CONTRIBUTING.md -->

<!-- Why am I raising this PR? Add context such as related issues, PRs, or documentation. -->
## Motivation

Add tags to EC2 Security Group resource when specified in CloudFormation template.

Fixes #10768 

<!-- What changes does this PR make? How does LocalStack behave differently now? -->
## Changes

Pass the `TagSpecifications` parameter to the `create_security_group` method when the `Tags` have been passed to the CloudFormation template.

<!-- Optional section: How to test these changes? -->

## Testing

Using this CloudFormation template:
```
Resources:
  SecurityGroup:
    Type: AWS::EC2::SecurityGroup
    Properties:
      GroupDescription: Security Group
      Tags:
        - Key: key1
          Value: value1
        - Key: key2
          Value: value2
```

run:
```
awslocal cloudformation create-stack --stack-name sg --template-body "$(cat cfn.yaml)"
awslocal ec2 describe-security-groups
```

Output should contain `Tags` in the newly created EC2 security group as in the following example:
```
        {
            "Description": "Security Group",
            "GroupName": "sg-SecurityGroup-4e42e960",
            "IpPermissions": [],
            "OwnerId": "000000000000",
            "GroupId": "sg-16639ba90ec152e26",
            "IpPermissionsEgress": [
                {
                    "IpProtocol": "-1",
                    "IpRanges": [
                        {
                            "CidrIp": "0.0.0.0/0"
                        }
                    ],
                    "Ipv6Ranges": [],
                    "PrefixListIds": [],
                    "UserIdGroupPairs": []
                }
            ],
            "Tags": [
                {
                    "Key": "key1",
                    "Value": "value1"
                },
                {
                    "Key": "key2",
                    "Value": "value2"
                }
            ],
            "VpcId": "vpc-e1f7c70b"
        }

```

<!-- Optional section: What's left to do before it can be merged? -->
<!--
## TODO

What's left to do:

- [ ] ...
- [ ] ...
-->

jcollado commented:
I have read the CLA Document and I hereby sign the CLA

pinzon reviewed (changes requested):
Great fix. However, to ensure that this change does not cause a regression, we should add an integration test or extend an existing one. I would recommend adding a test in the following file: `tests/aws/services/cloudformation/resource_providers/ec2/test_ec2.py`.

jcollado commented:
@pinzon Thanks for the pointer to the proper test file. That has been useful.

I've added a new test there that uses the same CloudFormation template that I used to reproduce the error and that I used to verify the same behavior in a real AWS account.

Despite it is not part of the issue, I'd like to highlight that AWS also adds the following tags to the resource:
- `aws:cloudformation:logical-id`
- `aws:cloudformation:stack-id`
- `aws:cloudformation:stack-name`

pinzon reviewed (changes requested):
Great job on the test! It just needs a few tweaks. At LocalStack, we're constantly working to improve compatibility with AWS. That's why we use Snapshots for validation – you can find more details about it [here](https://github.com/localstack/localstack/blob/master/docs/testing/parity-testing/README.md).

jcollado commented:
@pinzon Thanks for the information regarding snapshots. I've updated the test case to use a snapshot.

A few comments:

- There were some broken links in the documentation. I've updated them in this PR. If you prefer me to move those changes to a different PR, let me know.

- In the CloudFormation template I had to use `GetAtt` instead of `Ref` because AWS doesn't behave in the same way as LocalStack. From the documentation:

  > When you pass the logical ID of this resource to the intrinsic Ref function, Ref returns the ID of the security group if you specified the VpcId property. Otherwise, it returns the name of the security group.

  However, LocalStack always returns the ID of the security group. Should this behavior mismatch be reported in an issue?

- As commented above, CloudFormation adds additional tags to resources. For the test case to work, I had to filter them out from the response using custom code. Is there any filter transformer that can do the same job or is the custom code acceptable in this case?

pinzon commented on docs/testing/integration-tests/README.md:
Thanks!

pinzon commented on tests/aws/services/cloudformation/resource_providers/ec2/test_ec2.py:
```diff
@@ -56,6 +56,29 @@ def test_deploy_prefix_list(deploy_cfn_template, aws_client, snapshot):
+        )
+    )
+
+    snapshot.add_transformer(snapshot.transform.key_value("GroupId"))
+    snapshot.add_transformer(snapshot.transform.key_value("GroupName"))
+    snapshot.add_transformer(snapshot.transform.key_value("VpcId"))
+    snapshot.add_transformer(SortingTransformer("Tags", lambda tag: tag["Key"]))
+    response = aws_client.ec2.describe_security_groups(GroupIds=[stack.outputs["SecurityGroupId"]])
+    security_group = response["SecurityGroups"][0]
+
+    security_group["Tags"] = [
+        tag for tag in security_group["Tags"] if not tag["Key"].startswith("aws:cloudformation")
+    ]
```
Nice use of transformers.

The marker `@markers.snapshot.skip_verify` enables the skipping of certain attributes from validation using json paths, but it may not be applicable in this case. It's great that you filtered the tags that the resource provider, Another option could be for the resource provider to add them (even though it may be beyond the current scope of the issue).

pinzon commented on tests/aws/templates/ec2_security_group_with_tags.yml:
```diff
@@ -0,0 +1,14 @@
+
+Outputs:
+  SecurityGroupId:
+    Value: !GetAtt SecurityGroup.GroupId
```
We should address in a future PR the issue you mention about the Security Group Ref.

pinzon reviewed (approved):
Thank you for addressing my comments. I am happy to approve the PR. Let's proceed with merging it once all the necessary checks have passed.
